fix: renew an OIDC access token when the refresh token is not a JWT

kristinkomschow opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)
  - [ ] enable backports [when recommended](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)

## Related issues

closes #

kristinkomschow commented:
There is one issue that this will introduce (see https://github.com/camunda/camunda/issues/27710#issuecomment-2665923614):
> The downside of not checking refresh tokens is that the renew token request is still retried multiple times even when the error source is clear because of this code:
> 
> [camunda/operate/webapp/src/main/java/io/camunda/operate/webapp/security/identity/IdentityAuthentication.java](https://github.com/camunda/camunda/blob/63f7a76ae9ab1226466160e43ab24a1bca7b1c5c/operate/webapp/src/main/java/io/camunda/operate/webapp/security/identity/IdentityAuthentication.java#L283-L288)
> 
> Lines 283 to 288 in [63f7a76](/camunda/camunda/commit/63f7a76ae9ab1226466160e43ab24a1bca7b1c5c)
> 
>  private Tokens renewTokens(final String refreshToken) throws Exception { 
>    return getIdentityRetryService() 
>        .requestWithRetry( 
>            () -> getIdentity().authentication().renewToken(refreshToken), 
>            "IdentityAuthentication#renewTokens"); 
>  } 
> I could not find any justification for why this request is done with retries in the git history, so I'm unsure if we can just remove the retries on that request or if we should add better error handling.

I'd appreciate opinions on what I should do with this: remove retries, create new issue or adjust error handling (this may not be feasible because of time pressure)

sdorokhova commented on operate/webapp/src/main/java/io/camunda/operate/webapp/security/identity/IdentityAuthentication.java:
```diff
@@ -244,17 +214,6 @@ public void authenticate(final Tokens tokens) {
     subject = accessToken.getToken().getSubject();
     expires = accessToken.getToken().getExpiresAt();
     LOGGER.info("Access token will expire at {}", expires);
-    if (!isPolling()) {
```
It look like `isPolling` method can also be removed, right?

sdorokhova reviewed (approved):
LGTM! 
Just a tiny note in in `IdenityAuthentication` class.

kristinkomschow commented:
I removed the unused method.

I also added another change, after seeing this: https://learn.microsoft.com/en-us/entra/identity-platform/refresh-tokens#token-lifetime
According to the Microsoft Entra documentation, refresh tokens should be revoked once a new one has been fetched because this does not happen automatically. I added revoking the token in the refresh method.

kristinkomschow commented:
I added another change: retries for requests to Identity will only happen for `RestException` instead of all `IdentityException` classes now. This prevents retries to be executed when the issue is an expired/faulty token rather than a connection issue.
See discussion here: https://camunda.slack.com/archives/CRQAL5A1M/p1740053857648029?thread_ts=1739778227.737889&cid=CRQAL5A1M

Retries are only used for the token refresh and for getting authentication. I think it makes sense in both cases to only attempt a retry if the error isn't "just" a connection error.

sdorokhova reviewed (approved):
That's a nice catch with `IdentityException` vs `RestException`!

kristinkomschow commented:
Removed `refreshTokenExpiresAt ` field after Svetlana noted it and deemed it safe to delete despite session deserializatio that Spring does in `SessionService `:  "When persistent sessions are enabled, Spring will store sessions in Elasticsearch and read them whenever needed, this allows us to have several webapp nodes in SaaS. The user may be logged in on one node, but not on the other, but as sessions are persisted they can be reused. I'm actually sure now that it's safe to delete the field. I checked the code and if session can't be serialized Spring will simply consider that there is no session yet and will create the new one."
